Validate issued_at field to prevent future dates further than 1 year

tatheerf02 opened the pull request:
Resolves #4742

### Description

Add a new validation to the `issued_at` field which is present on the donation, distribution, and purchase models. The new validation ensures that the field can not be set to a date that is further than one year into the future, as this causes unexpected behaviour when the models are later filtered.

An alternative solution could be to add this validation to the models themselves, but putting it in the centralized location was an existing pattern and made more sense.

Tradeoff: Are there valid use cases where we might want to have set this date to further than one year ahead?

### Type of change

* Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?

Added new unit tests following existing patterns for all affected models and ensured they passed

### Screenshots

**Note**: Although not pictured, I also tested that the validation is run when the date on a purchase, donation, or distribution is edited after creation.

<details><summary>📸 Validation error shown on new donation</summary>
<img width="1085" alt="Screen Shot 2024-10-28 at 9 59 30 PM" src="https://github.com/user-attachments/assets/8ff5c4e7-b08d-44ec-a8ed-28be7f863f78">
</details>

<details><summary>📸 Validation error shown on new purchase</summary>
<img width="1085" alt="Screen Shot 2024-10-28 at 9 59 00 PM" src="https://github.com/user-attachments/assets/229bf122-2888-48c5-b302-1bd7f8484981">
</details>

<details><summary>📸 Validation error shown on new distribution</summary>
<img width="1085" alt="Screen Shot 2024-10-28 at 9 58 41 PM" src="https://github.com/user-attachments/assets/e73efa96-650d-435c-b4ce-acb4d61dfbbf">
<details>

cielf commented:
@tatheerf02 Let's change that from "Cannot be further than 1 year" to "cannot be more than 1 year in the future".

tatheerf02 commented:
> Let's change that from "Cannot be further than 1 year" to "cannot be more than 1 year in the future".

@cielf I've changed the error message and updated the screenshots in the description 👍

cielf reviewed (commented):
LGTM from a functional pov; I found an unrelated issue (which is also in main).

cielf reviewed (commented):
I took a glancing look at the code, and I'm not sure all the changes made to the calendar_service_spec are in scope?  Passing it to @awwaiid or @dorner for an opinion, and other technical review.

dorner reviewed (approved):
Looks good. The calendar spec had to change because it had a year too far in the future, which would prevent the object from saving now.
